### PR TITLE
Improve nlu machine learning infra

### PR DESCRIPTION
**Pull request by ruinianxu**
# Description

This PR is mainly for improving NLU retraining pipeline which automatically runs checksum computation and file zip and upload. 

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

Currently, this script supports training a single job in the local machine and automatically storing the best model and uploading it to the AWS. 

The next step is to incorporate parameter sweep and training via SLRUM. 

# Testing
For the purpose of testing, please use batch id of 20220524_1653470045. Before running retraining script, you will need to run the following script to generate split mask of the old data. We consider all data to be new here. Move the generated split_masks.pth under /fairo/droidlet/tools/hitl/nsp_retrain/configs.
```
import torch
# start and end index for train, val and test split
train_s_idx = 0
train_e_idx = 0
val_s_idx = 0
val_e_idx = 0
test_s_idx = 0
test_e_idx = 0
# generate list
train_list = [i for i in range(train_s_idx, train_e_idx)]
val_list = [i for i in range(val_s_idx, val_e_idx)]
test_list = [i for i in range(test_s_idx, test_e_idx)]
# turn into tensor
train_list = torch.tensor(train_list)
val_list = torch.tensor(val_list)
test_list = torch.tensor(test_list)
save_path = "./split_masks.pth"
data = {"annotated": {
"train": train_list,
"valid": val_list,
"test": test_list,
}}
torch.save(data, save_path)
```
Then run
```
python droidlet/tools/hitl/nsp_retrain_infra.py --batch_id 20220524_1653470045 --retrain_data_splits 2 2 2 --num_epochs 2 -decoder_learning_rate .0000005 .000001 .000005 -encoder_learning_rate .0 .000001
```

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [x] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I ran on hardware (1) all scripts in `tests/scripts`, (2) asv benchmarks.

### CODE REVIEW

File: droidlet/perception/semantic_parsing/nsp_transformer_model/train_model.py [hunk 1]
```diff
@@ -269,13 +385,16 @@ def eval_model_on_dataset(
         l, _, _, a, text_span_acc, text_span_loss = self.validate(
             encoder_decoder, valid_dataset, tokenizer, self.args
         )
-        print("evaluating on {} valid: \t Loss: {:.4f} \t Accuracy: {:.4f}".format(dtype, l, a))
-        logging.info(
-            "evaluating on {} valid: \t Loss: {:.4f} \t Accuracy: {:.4f}".format(dtype, l, a)
-        )
+        # logging.info(
```
Comment (kavyasrinet) [inline droidlet/perception/semantic_parsing/nsp_transformer_model/train_model.py:388]: nit: can you remove the comment ?

File: droidlet/perception/semantic_parsing/nsp_transformer_model/train_model_clean.py [hunk 2]
```diff
@@ -0,0 +1,737 @@
+"""
+Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (kavyasrinet) [inline droidlet/perception/semantic_parsing/nsp_transformer_model/train_model_clean.py:2]: why do we need a new file ? can we name this better to reflect the purpose of the file ?

File: droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py [hunk 3]
```diff
@@ -14,6 +14,72 @@
 print("Rootdir : %r" % ROOTDIR)
 
 
+def tar_and_upload(agent, artifact_name, model_name=None):
```
Comment (kavyasrinet) [inline droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py:17]: I don't understand why we need this new function ? The function : `compute_checksum_tar_and_upload` does exactly this right ?

File: droidlet/tools/hitl/nsp_retrain/scripts/sweep_runner.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        g.write("# opts :: " + varying_args[i] + " :: " + outpaths[i] + "\n")
+    for i in range(len(all_arglists)):
+        g.write("echo " + "'" + varying_args[i] + " :: " + outpaths[i] + "'" + "\n")
+    for i in range(len(all_arglists)):
+        g.write("sbatch " + str(i) + ".sh &\n")
+    g.close()
+    st = os.stat(mastername)
+    os.chdir(scripts_dir)
+    os.chmod(mastername, st.st_mode | stat.S_IEXEC)
+    os.system(mastername)
```
Comment (kavyasrinet) [inline droidlet/tools/hitl/nsp_retrain/scripts/sweep_runner.py:211]: Thanks for adding this file in. Can you adda README to the `nsp_retrain` folder highlighting the purpose of the new files ?

File: droidlet/tools/hitl/nsp_retrain/scripts/sweep_runner.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+    # Moving sweep monitoring to be internal to nsp_retrain_infra.py
+
+    # this_dir = os.path.dirname(os.path.abspath(__file__))
+    # shutil.copyfile(os.path.join(this_dir, "sweep_monitor.py"), os.path.join(model_out, "sweep_monitor.py"))
+    # for i in range(300):
+    #    cmd = 'echo -e "cd {}\n python sweep_monitor.py" | at now +{} minutes'.format(model_out, (i+1)*10)
```
Comment (kavyasrinet) [inline droidlet/tools/hitl/nsp_retrain/scripts/sweep_runner.py:218]: nit: do we need to check in the commented section here ?

File: droidlet/tools/hitl/nsp_retrain_infra.py [hunk 6]
```diff
@@ -51,40 +54,48 @@ def max_and_argmax(self, l):
         i = max(range(len(l)), key=lambda i: l[i])
         return l[i], i
 
-    def get_accs(self):
+    def get_accs_losses(self):
```
Comment (kavyasrinet) [inline droidlet/tools/hitl/nsp_retrain_infra.py:57]: nit: rename to `get_accuracy_and_losses`

File: tools/nsp_scripts/test_model_script.py [hunk 7]
```diff
@@ -15,31 +22,258 @@
...
+        logging.info("====== Making Grammar ======")
+        build_grammar(args)
+    with open(args.tree_voc_file) as fd:
+        full_tree, tree_i2w = json.load(fd)
+    full_tree_voc = (full_tree, tree_i2w)
+
+    logging.info("====== Loading Pretrained Parameters ======")
+    sd, _, _, _, _ = load_model(args.model_dir)
+    logging.info("====== Setting up Model ======")
+    decoder_with_loss, encoder_decoder, tokenizer = build_model(args, full_tree_voc[1])
```
Comment (kavyasrinet) [inline tools/nsp_scripts/test_model_script.py:263]: nit: this looks like a file from a different PR ?

**Review by kavyasrinet — COMMENTED** [review]
Can you rebase on main. We shouldn't have this PR make any changes to the file: `polymetis/polymetis/src/clients/franka_panda_client/third_party/libfranka`

File: droidlet/perception/semantic_parsing/nsp_transformer_model/train_model_clean.py (hunk 2 above)
Comment (ruinianxu) [reply droidlet/perception/semantic_parsing/nsp_transformer_model/train_model_clean.py:2]: @kavyasrinet 
I was not sure if I should clean up the previous training script without making sure that new retraining pipeline is working. I named train_model_clean just to indicate it this is the cleaned-up version of training script. If it looks fine, I will replace the previous one with this one and remove it.

File: droidlet/perception/semantic_parsing/nsp_transformer_model/train_model.py (hunk 1 above)
Comment (ruinianxu) [reply droidlet/perception/semantic_parsing/nsp_transformer_model/train_model.py:388]: @kavyasrinet sure.

File: tools/nsp_scripts/test_model_script.py (hunk 7 above)
Comment (ruinianxu) [reply tools/nsp_scripts/test_model_script.py:263]: @kavyasrinet Yeah. It seems somehow I updated this script with the change in this [PR](https://github.com/facebookresearch/fairo/pull/1201). Should I change it back to the script of main branch?

Comment (kavyasrinet) [reply tools/nsp_scripts/test_model_script.py:263]: We should merge your other PR first. Once that's merged in and we rebase this PR on main, this should go away. But ideally yes, every PR should be against main branch and shouldn't have changes from the other PRs

Comment (ruinianxu) [reply tools/nsp_scripts/test_model_script.py:263]: Gotcha.

**Comment by ruinianxu** [comment]
> Can you rebase on main. We shouldn't have this PR make any changes to the file: `polymetis/polymetis/src/clients/franka_panda_client/third_party/libfranka`

Hi Kavya, I am not sure why there are some changes committed to this sub module. Previously, I created a new branch for code clean up and rebase it to the main branch, which leads to the change of this submodule.  It seems it didn't automatically checkout the specific commit. Then I ran ```git submodule update``` and it checkout to the right commit. 

Not sure what to do for this PR since rebase seems not resolve the issue. Maybe run ```git submodule update``` and commit the update? Thanks for help!

**Comment by kavyasrinet** [comment]
> > Can you rebase on main. We shouldn't have this PR make any changes to the file: `polymetis/polymetis/src/clients/franka_panda_client/third_party/libfranka`
> 
> Hi Kavya, I am not sure why there are some changes committed to this sub module. Early this afternoon, I created a new branch for code clean up and rebase it to the main branch, which leads to the same change of this sub module. It seems it didn't automatically checkout the specific commit. Then I ran `git submodule update` and it checkout to the right commit.
> 
> Not sure what to do for this PR since rebase seems not resolve the issue. Maybe run `git submodule update` and commit the update? Thanks for help!

Have you tried that on this branch ?
